File: app/src/main/java/com/blueskybone/arkscreen/base/preference/PreferenceStore.java
package com.blueskybone.arkscreen.base.preference;

/*
 * The code for data persistence refers to the following project
 * https://github.com/easybangumiorg/EasyBangumi
 * */

import java.util.Set;
import java.util.function.Function;

public interface PreferenceStore {
    Preference<String> getString(String key, String value);

    Preference<Integer> getInt(String key, int value);

    Preference<Long> getLong(String key, Long value);

    Preference<Float> getFloat(String key, Float value);

    Preference<Boolean> getBoolean(String key, Boolean value);

    Preference<Set<String>> getStringSet(String key, Set<String> set);

    <T> Preference<T> getObject(
            String key,
            T defaultValue,
            Function<T, String> serializer,
            Function<String, T> deserializer
    );

    default <T extends Enum<T>> Function<T, String> serializer() {
        return T::name;
    }

    default <T extends Enum<T>> Function<String, T> deserializer(T defaultValue) {
        return string -> {
            try {
                return Enum.valueOf((Class<T>) defaultValue.getDeclaringClass(), string);
            } catch (Exception e) {
                return defaultValue;
            }
        };
    }

    default <T extends Enum<T>> Preference<T> getEnum(String key, T defaultValue) {
        return getObject(key, defaultValue, serializer(), deserializer(defaultValue));
    }
}

File: app/src/main/java/com/blueskybone/arkscreen/preference/preference/PreferenceStore.kt
package com.blueskybone.arkscreen.preference.preference

import java.util.function.Function

/**
 *   Created by blueskybone
 *   Date: 2025/1/7
 */
interface PreferenceStore {
    fun getString(key: String, value: String): Preference<String>

    fun getInt(key: String, value: Int): Preference<Int>

    fun getLong(key: String, value: Long): Preference<Long>

    fun getFloat(key: String, value: Float): Preference<Float>

    fun getBoolean(key: String, value: Boolean): Preference<Boolean>

    fun <T> getObject(
        key: String,
        defaultValue: T,
        serializer: Function<T, String>,
        deserializer: Function<String, T>
    ): Preference<T>
}